cursive/src/backends/wasm.rs

```diff
@@ -0,0 +1,266 @@
+#![cfg(feature = "wasm-backend")]
+
+use cursive_core::{
+    event::{ Event, Key },
+    Vec2,
+    theme,
+};
+use std::collections::VecDeque;
+use std::rc::Rc;
+use std::cell::RefCell;
+use web_sys::HtmlCanvasElement;
+use wasm_bindgen::prelude::*;
+use crate::backend;
+
+#[wasm_bindgen]
+#[derive(Debug, PartialEq)]
+#[repr(C)]
+struct TextColorPair  {
+    text: char,
+    color: ColorPair,
+}
+
+impl TextColorPair {
+    pub fn new(text: char, color: ColorPair) -> Self {
+        Self {
+            text,
+            color,
+        }
+    }
+}
+
+fn text_color_pairs_to_bytes(buffer: &Vec<TextColorPair>) -> &[u8] {
+    unsafe {
+        std::slice::from_raw_parts(
+            buffer.as_ptr() as *const u8,
+            buffer.len() * std::mem::size_of::<TextColorPair>(),
+        )
+    }
+}
+
+impl Clone for TextColorPair {
+    fn clone(&self) -> Self {
+        Self {
+            text: self.text,
+            color: self.color.clone(),
+        }
+    }
+}
+
+
+#[wasm_bindgen(module = "/src/backends/canvas.js")]
+extern "C" {
+    fn paint(buffer: &[u8]);
+}
+
+/// Backend using wasm.
+pub struct Backend {
+    canvas: HtmlCanvasElement,
+    width: usize,
+    height: usize,
+    color: RefCell<ColorPair>,
+    events: Rc<RefCell<VecDeque<Event>>>,
+    buffer: RefCell<Vec<TextColorPair>>,
+}
+impl Backend {
+    /// Creates a new Cursive root using a wasm backend and given HTML canvas.
+    pub fn new(canvas: HtmlCanvasElement) -> std::io::Result<Box<dyn backend::Backend>> {
+        let document = web_sys::window()
+            .ok_or(std::io::Error::new(
+                std::io::ErrorKind::Other,
+                "Failed to get window",
+            ))?
+            .document()
+            .ok_or(std::io::Error::new(
+                std::io::ErrorKind::Other,
+                "Failed to get document",
+            ))?;
+
+        let color = cursive_to_color_pair(theme::ColorPair {
+            front: theme::Color::Light(theme::BaseColor::Black),
+            back:theme::Color::Dark(theme::BaseColor::Green),
+        });
+
+        let events = Rc::new(RefCell::new(VecDeque::new()));
+        let cloned = events.clone();
+        let closure = Closure::wrap(Box::new(move |event: web_sys::KeyboardEvent| {
+            match event.key_code() {
+                8 => cloned.borrow_mut().push_back(Event::Key(Key::Backspace)),
+                13 => cloned.borrow_mut().push_back(Event::Key(Key::Enter)),
+                37 => cloned.borrow_mut().push_back(Event::Key(Key::Left)),
+                38 => cloned.borrow_mut().push_back(Event::Key(Key::Up)),
+                39 => cloned.borrow_mut().push_back(Event::Key(Key::Right)),
+                40 => cloned.borrow_mut().push_back(Event::Key(Key::Down)),
+                code => {
+                    if let Some(c) = std::char::from_u32(code) {
+                        cloned.borrow_mut().push_back(Event::Char(c));
+                    }
+                }            
```

Comment:
What about all the other key-events, such as Escape, control/alt keys, etc? These are all also retrievable, going by the [`web_sys::KeyboardEvent`](https://docs.rs/web-sys/latest/web_sys/struct.KeyboardEvent.html) documentation.

Should be implemented too for completeness sake - as well as control-key-sequences are rather often used by TUI applications.

Also, rewriting/simplifying this to something like this form would be good readability-wise:
```suggestion
            let event = match event.key_code() {
                8 => Event::Key(Key::Backspace),
                ..
             };
             cloned.borrow_mut().push_back(event);
```